MariaDB/server
# MDEV-11412: Backporting to 10.2 and small adjustments

- [x] *The Jira issue number for this PR is: MDEV-11412*


## Description
As a user, I expect that a table will be fully dropped from the database when I issue a `DROP TABLE` statement, no matter of it's underlying state. However in case of absent .frm file MariaDB couldn't determine a storage engine for the table.
This problem was documented in [MDEV-11412](https://jira.mariadb.org/browse/MDEV-11412) and the fix was released in 10.5.4.

This PR is backports the fix from 10.5 to earlier 10.2.
The following commits are cherry-picked from 10.5:
- 5bcb1d65
- 346d10a9
- 4080e3ac
- dfb41fdd

In addition there is a commit adapting the code for 10.2 specifically.

## How can this PR be tested?

The most impactful change is related to `mysql_rm_table_no_locks` method. Please note that this change was already shipped with MariaDB 10.5.4 and it has not caused any known regressions so far.

A good testing scenario was previously introduced in `drop_table_force.test`. It was adjusted along with 22 other MTR tests to reflect the current state of the engine, which provides a good test coverage for this change.


## Basing the PR against the correct MariaDB version

Please note that similar patches for 10.3 and 10.4 have already been prepared. I am submitting this for 10.2 first to get feedback/acknowledgement before I submit the PRs for 10.3 and 10.4.

Since 10.3 introduces `SEQUENCE`, some code was removed from the original commits being backported to 10.2, which means the implementation as I'm suggesting cannot be directly merged to a later major version.

If MariaDB maintainers are willing to accept the change please keep that in mind and let me know what is the preferred course of action in order to get it delivered to 10.2-10.4. If you prefer, I can submit the PR for 10.4 and 10.3 so those can be merged before this 10.2 version.

## Copyright

All new code of the whole pull request, including one or several files that are either new files or modified ones, are contributed under the BSD-new license. I am contributing on behalf of my employer Amazon Web Services, Inc.
